akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -253,9 +253,11 @@ object ByteString {
       else ByteString1(bytes, startIndex, Math.min(n, length))
 
     override def slice(from: Int, until: Int): ByteString = {
-      if (from <= 0 && until >= length) this // we can do < / > since we're Compact
+      if (from <= 0 && length <= until) this // we can do < / > since we're Compact
       else if (until <= from) ByteString1.empty
-      else ByteString1(bytes, startIndex + from, until - from)
+      else if (0 <= from && until <= bytes.length) ByteString1(bytes, startIndex + from, until - from)
+      else if (from < 0) ByteString1(bytes, startIndex, until)
+      else /*(until > bytes.length)*/ ByteString1(bytes, startIndex + from, bytes.length - from)
     }
```

Comment:
You're right, my snippet blows up on `slice(10, 1000)`. I need to look more carefully... 
Trying to minimise if branches, slowdown was noticeable hm, if we'd stay around 250M/s that'd be good - you can check in the akka-bench-jmh benchmarks